File: backend/src/main/java/backend/service/ReviewService.java
package backend.service;

import backend.dto.request.ReviewCreateRequest;
import backend.dto.response.ReviewResponse;
import backend.entity.Review;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ReviewService {
    
    /**
     * Get all approved reviews
     */
    List<ReviewResponse> getAllApprovedReviews();
    
    /**
     * Create a new review (User only - after completing tour)
     */
    ReviewResponse createReview(ReviewCreateRequest request, Long userId);
    
    /**
     * Update own review (User only)
     */
    ReviewResponse updateReview(Long reviewId, ReviewCreateRequest request, Long userId);
    
    /**
     * Delete own review (User only)
     */
    void deleteReview(Long reviewId, Long userId);
    
    /**
     * Get review by ID
     */
    ReviewResponse getReviewById(Long reviewId);
    
    /**
     * Get reviews by tour ID (approved only for public)
     */
    List<ReviewResponse> getReviewsByTourId(Long tourId);
    
    /**
     * Get reviews by tour ID with pagination
     */
    Page<ReviewResponse> getReviewsByTourId(Long tourId, Pageable pageable);
    
    /**
     * Get reviews by user ID
     */
    List<ReviewResponse> getReviewsByUserId(Long userId);
    
    /**
     * Get reviews by user ID with pagination
     */
    Page<ReviewResponse> getReviewsByUserId(Long userId, Pageable pageable);
    
    /**
     * Vote review as helpful
     */
    ReviewResponse voteHelpful(Long reviewId, Long userId);
    
    /**
     * Check if user has already reviewed a tour
     */
    boolean hasUserReviewedTour(Long userId, Long tourId);
    
    /**
     * Check if user can review (booking exists and completed)
     */
    boolean canUserReviewTour(Long userId, Long tourId, Long bookingId);
    
    /**
     * Check if user has completed booking for a tour (without specific bookingId)
     */
    boolean hasCompletedBookingForTour(Long userId, Long tourId);
    
    /**
     * Get AI-generated summary of reviews for a tour
     */
    backend.dto.response.ReviewAiSummaryResponse getAiSummary(Long tourId);
    
    /**
     * Get review by booking ID
     */
    ReviewResponse getReviewByBookingId(Long bookingId);
    
    /**
     * Calculate average rating for tour
     */
    Double calculateAverageRating(Long tourId);
    
    /**
     * Get rating distribution for tour
     */
    java.util.Map<Integer, Long> getRatingDistribution(Long tourId);
    
    // ========== ADMIN METHODS ==========
    
    /**
     * Get all reviews (Admin - includes all statuses)
     */
    Page<ReviewResponse> getAllReviewsAdmin(Pageable pageable);
    
    /**
     * Get reviews by status (Admin)
     */
    Page<ReviewResponse> getReviewsByStatus(String status, Pageable pageable);
    
    /**
     * Approve review (Admin)
     */
    ReviewResponse approveReview(Long reviewId);
    
    /**
     * Reject review (Admin)
     */
    ReviewResponse rejectReview(Long reviewId, String reason);
    
    /**
     * Update review status (Admin)
     */
    ReviewResponse updateReviewStatus(Long reviewId, String status);
    
    /**
     * Reply to review (Admin)
     */
    ReviewResponse replyToReview(Long reviewId, String reply);
    
    /**
     * Delete reply from review (Admin)
     */
    ReviewResponse deleteReply(Long reviewId);
    
    /**
     * Delete any review (Admin)
     */
    void deleteReviewAdmin(Long reviewId);
    
    /**
     * Get total reviews count
     */
    long getTotalReviews();
    
    /**
     * Get pending reviews count
     */
    long getPendingReviewsCount();
    
    /**
     * Get reviews with advanced filters (Admin)
     */
    Page<ReviewResponse> getReviewsWithFilters(
        String status, Long tourId, Integer rating,
        java.time.LocalDateTime startDate, java.time.LocalDateTime endDate,
        Boolean isSuspicious, Boolean isSpam, Pageable pageable
    );
    
    /**
     * Mark review as suspicious
     */
    ReviewResponse markAsSuspicious(Long reviewId, boolean suspicious);
    
    /**
     * Mark review as spam
     */
    ReviewResponse markAsSpam(Long reviewId, boolean spam);
    
    /**
     * Analyze review with AI (detect negative keywords, sentiment)
     */
    java.util.Map<String, Object> analyzeReviewWithAI(Long reviewId);
    
    /**
     * Get all reviews for export (with filters)
     */
    java.util.List<ReviewResponse> getAllReviewsForExport(
        String status, Long tourId, Integer rating,
        java.time.LocalDateTime startDate, java.time.LocalDateTime endDate,
        Boolean isSuspicious, Boolean isSpam
    );
}

File: frontend/src/services/reviewService.ts
import api from './api';

export interface ReviewCreateRequest {
  tourId: number;
  bookingId: number;
  rating: number;
  comment: string;
}

export interface ReviewUpdateRequest {
  rating: number;
  comment: string;
}

export interface UserSummary {
  id: number;
  name: string;
  avatarUrl?: string;
}

export interface TourSummary {
  id: number;
  name: string;
  slug: string;
  mainImage?: string;
}

export interface ReviewResponse {
  id: number;
  rating: number;
  comment?: string;
  status: string;
  helpfulCount: number;
  images?: string[];
  createdAt: string;
  updatedAt: string;
  adminReply?: string;
  repliedAt?: string;
  repliedBy?: number;
  user: UserSummary;
  tour: TourSummary;
  bookingId: number;
}

export interface RatingDistribution {
  [key: number]: number;
}

export interface TourRatingStats {
  averageRating: number;
  ratingDistribution: RatingDistribution;
}

const reviewService = {
  /**
   * Create a new review
   */
  createReview: async (data: ReviewCreateRequest): Promise<ReviewResponse> => {
    const response = await api.post('/reviews', data);
    return response.data.data;
  },

  /**
   * Update own review
   */
  updateReview: async (reviewId: number, data: ReviewUpdateRequest): Promise<ReviewResponse> => {
    const response = await api.put(`/api/reviews/${reviewId}`, data);
    return response.data.data;
  },

  /**
   * Delete own review
   */
  deleteReview: async (reviewId: number): Promise<void> => {
    await api.delete(`/api/reviews/${reviewId}`);
  },

  /**
   * Get review by ID
   */
  getReviewById: async (reviewId: number): Promise<ReviewResponse> => {
    const response = await api.get(`/api/reviews/${reviewId}`);
    return response.data.data;
  },

  /**
   * Get reviews by tour ID
   */
  getReviewsByTourId: async (tourId: number): Promise<ReviewResponse[]> => {
    const response = await api.get(`/api/reviews/tour/${tourId}`);
    return response.data.data;
  },

  /**
   * Get reviews by tour ID with pagination
   */
  getReviewsByTourIdPaginated: async (
    tourId: number,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'createdAt',
    sortDirection: string = 'desc'
  ): Promise<{
    content: ReviewResponse[];
    totalElements: number;
    totalPages: number;
    pageNumber: number;
    pageSize: number;
    first: boolean;
    last: boolean;
    empty: boolean;
  }> => {
    const response = await api.get(`/api/reviews/tour/${tourId}/paginated`, {
      params: { page, size, sortBy, sortDirection }
    });
    return response.data.data;
  },

  /**
   * Get current user's reviews
   */
  getMyReviews: async (): Promise<ReviewResponse[]> => {
    const response = await api.get('/reviews/my-reviews');
    return response.data.data;
  },

  /**
   * Get reviews by user ID
   */
  getReviewsByUserId: async (userId: number): Promise<ReviewResponse[]> => {
    const response = await api.get(`/api/reviews/user/${userId}`);
    return response.data.data;
  },

  /**
   * Vote a review as helpful
   */
  voteHelpful: async (reviewId: number): Promise<ReviewResponse> => {
    const response = await api.post(`/api/reviews/${reviewId}/helpful`);
    return response.data.data;
  },

  /**
   * Check if user can review a tour
   */
  canReviewTour: async (tourId: number, bookingId: number): Promise<boolean> => {
    const response = await api.get('/reviews/can-review', {
      params: { tourId, bookingId }
    });
    return response.data.data.canReview;
  },

  /**
   * Get review by booking ID
   */
  getReviewByBookingId: async (bookingId: number): Promise<ReviewResponse | null> => {
    const response = await api.get(`/api/reviews/booking/${bookingId}`);
    return response.data.data;
  },

  /**
   * Get tour rating statistics
   */
  getTourRatingStats: async (tourId: number): Promise<TourRatingStats> => {
    const response = await api.get(`/api/reviews/tour/${tourId}/stats`);
    return response.data.data;
  },
};

export default reviewService;

